include/swift/AST/Expr.h

```diff
@@ -3678,6 +3687,8 @@ class ClosureExpr : public AbstractClosureExpr {
   /// Is this a completely empty closure?
   bool hasEmptyBody() const;
 
+  VarDecl *getCapturedSelfDecl() const { return CapturedSelfDecl; }
```

Comment:
Rather than explicitly storing `CapturedSelfDecl`, consider looping over the `CaptureListEntry`s here and returning any that returns true to `isSimpleSelfCapture()`. 

That would simplify the initializers and parsing methods (one less parameter to pass around), and also simplify the diagnostic logic, since this would only be non-null if it is a simple self capture, so you wouldn't need to check both things.

Reply:
The issue with that approach is that the `CaptureListEntry` array is only available on `CaptureListExpr` (which is a parent of the `ClosureExpr` in the AST), but the `ClosureExpr` is the `DeclContext` that we have available at lookup time. 

A couple commits back you can see the implementation which added a backedge to the AST so that we could recover a `CaptureListExpr` from a `ClosureExpr` when needed, but based on some feedback from @brentdax I pursued this direction instead. Given that we *do* have the information needed at parse-time I tend to prefer this solution since it doesn’t muddy the AST, even though it comes at the cost of slightly more complex parsing and diagnostic logic.

If you have further thoughts on the trade off here, though, I’m definitely open to suggestions!

Reply:
Ah! No, that makes perfect sense. Storing extra is definitely better than extra AST back pointers. Sorry I hadn't looked back at the history.